ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/LagReportingAgent.java

```diff
@@ -0,0 +1,380 @@
+    }
+  }
+
+  /**
+   * Stores the host lag received from a remote Ksql server.
+   * @param lagReportingRequest The host lag information sent directly from the other node.
+   */
+  public void receiveHostLag(final LagReportingRequest lagReportingRequest) {
+    final long updateTimeMs = lagReportingRequest.getLastLagUpdateMs();
+    final HostInfoEntity hostInfoEntity = lagReportingRequest.getHostInfo();
+    final HostInfo hostInfo = new HostInfo(hostInfoEntity.getHost(), hostInfoEntity.getPort());
+
+    ImmutableMap.Builder<LagInfoKey, Map<Integer, LagInfoEntity>> hostMapBuilder
+        = ImmutableMap.builder();
+    for (Map.Entry<String, Map<Integer, LagInfoEntity>> storeEntry
+        : lagReportingRequest.getStoreToPartitionToLagMap().entrySet()) {
+      final LagInfoKey lagInfoKey = LagInfoKey.of(storeEntry.getKey());
+      final Map<Integer, LagInfoEntity> partitionMap = storeEntry.getValue();
+
+      ImmutableMap.Builder<Integer, LagInfoEntity> partitionsBuilder = ImmutableMap.builder();
+
+      // Go through each new partition and add lag info
+      for (final Map.Entry<Integer, LagInfoEntity> partitionEntry : partitionMap.entrySet()) {
+        final Integer partition = partitionEntry.getKey();
+        final LagInfoEntity lagInfo = partitionEntry.getValue();
+        partitionsBuilder.put(partition, lagInfo);
+      }
+
+      hostMapBuilder.put(lagInfoKey, partitionsBuilder.build());
+    }
+
+    Map<LagInfoKey, Map<Integer, LagInfoEntity>> hostMap = hostMapBuilder.build();
+    HostLagInfo hostLagInfo = new HostLagInfo(hostMap, updateTimeMs);
+    receivedLagInfo.compute(hostInfo, (hi, previousHostLagInfo) ->
+        previousHostLagInfo != null && previousHostLagInfo.getUpdateTimeMs() > updateTimeMs ?
+            previousHostLagInfo : hostLagInfo);
+  }
+
+  /**
+   * Returns lag information for all of the "alive" hosts for a given state store and partition.
+   * @param lagInfoKey The lag info key
+   * @param partition The partition of that state
+   * @return A map which is keyed by host and contains lag information
+   */
+  public Map<HostInfo, LagInfoEntity> getHostsPartitionLagInfo(
+      final Set<HostInfo> hosts, final LagInfoKey lagInfoKey, final int partition) {
+    ImmutableMap.Builder<HostInfo, LagInfoEntity> builder = ImmutableMap.builder();
+    for (HostInfo host : hosts) {
+      LagInfoEntity lagInfo = receivedLagInfo.getOrDefault(host, EMPTY_HOST_LAG_INFO).getLagInfo()
+          .getOrDefault(lagInfoKey, Collections.emptyMap())
+          .getOrDefault(partition, null);
+      if (aliveHosts.containsKey(host) && lagInfo != null) {
+        builder.put(host, lagInfo);
+      }
+    }
+    return builder.build();
+  }
+
+  /**
+   * Returns a map of storeName -> partition -> LagInfoEntity.  Meant for being exposed in testing
+   * and debug resources.
+   */
+  public Map<String, Map<String, Map<Integer, LagInfoEntity>>> listAllLags() {
+      return receivedLagInfo.entrySet().stream()
```

Comment:
this block of code, I find it hard to read.. Do you think we can just have a class `TablePartitionLagInfo` which wraps a `queryId`, `storeName`, `partition`, `lagInfo` and simply manage a list/map tht is just one level? I went with such a nested map because thats how Streams code was.. but I think we can add an abstraction here, to make this easier?

Reply:
To make it a single level which would be usable by Jackson (which this is since it's exposed in an endpoint) would require "stringifying" the host, storename, partition as a single key and would make it harder to pick apart 
in tests.  I'll instead use Jackson-friendly types rather than Strings if that would make it easier to read.

I'm fine to rewrite this as a normal block without using so many streams.  I went a little overboard :-)